packages/framework/src/Model/Product/ProductVisibilityRepository.php

```diff
@@ -330,4 +341,48 @@ protected function hideMainVariantsWithoutVisibleVariants($onlyMarkedProducts)
             ]
         );
     }
+
+    protected function refreshEntityState(): void
+    {
+        $productClassName = $this->entityNameResolver->resolve(Product::class);
+        $productVisibilitiesClassName = $this->entityNameResolver->resolve(ProductVisibility::class);
+
+        $this->refreshAllEntitiesByClassName($productClassName);
+        $this->refreshAllEntitiesByClassName($productVisibilitiesClassName);
+    }
+
+    /**
+     * @param string $className
+     * @internal
+     */
+    protected function refreshAllEntitiesByClassName(string $className): void
+    {
+        $identityMap = $this->em->getUnitOfWork()->getIdentityMap();
+
+        if (!array_key_exists($className, $identityMap)) {
+            return;
+        }
+
+        foreach ($identityMap[$className] as $entity) {
+            $this->em->refresh($entity);
+        }
+    }
```

Comment:
I think this method should be moved to `EntityManagerDecorator` including resolving entity names

Reply:
Cool idea, thanks!
Done.